File: packages/slot-math-kernels/src/slot_math_kernels/money_collect.py
"""W244 wave 10 — closed-form analytical model for `money_collect` feature.

Industry pattern (Cash Eruption / BTG Money Train / Pragmatic Coin Volcano):

  Trigger
  -------
    Initial spin lands ≥ trigger_count_min money symbols (and any non-money
    paid wins are evaluated as normal). Triggering re-flips reels into a
    "cash bonus" mode where:

      * Money symbols already on the grid LOCK in place with their value.
      * `money_respins_reset` (typical 3) respins are awarded.
      * Each subsequent respin: every new money symbol that lands also LOCKS
        and RESETS the respin counter to `money_respins_reset`.
      * Episode ends when:
          (a) `money_respins_reset` consecutive spins land no new money
              symbols, OR
          (b) the grid fills up (`money_grid_cap` total locked symbols).

  Payout
  ------
    Final award = SUM(value_x_bet over all locked symbols) × bet.
    Each money symbol's value drawn IID from `money_value_weights`
    distribution (a {value × bet : weight} table — e.g. {1:50, 2:30, 5:15,
    10:5} normalised internally).

Closed-form RTP component
-------------------------
This module computes the *expected* per-spin money_collect contribution
to total RTP, given:

  * `p_per_cell` — probability a single grid cell shows money on a respin
    (estimated from money symbol reel-weight share).
  * `n_cells`   — grid size (e.g. 15 for 5×3, 24 for 6×4, 20 for ways).
  * `trigger_p` — initial-spin trigger probability (≥ trigger_count_min
    money symbols out of `n_cells` Bernoulli trials).
  * `value_table` — normalised {value_x_bet : prob}.
  * `respins_reset` — typical 3.
  * `grid_cap` — full-grid stopping rule.

  Per-spin RTP (money_collect component) =
      trigger_p × E[total_money_value_per_episode]

  where E[total_money_value_per_episode] is a Markov chain over
  (k_locked, respins_remaining) — analytic via dynamic-programming on the
  state space.

The DP runs in O(grid_cap × respins_reset) state transitions which is
typically < 100 — sub-millisecond regardless of `n_cells`.

This module is pure-stdlib and used by:
  * `tools.math_dsl.compile` — emits IR with `meta.rtp_breakdown.money_collect`
  * `tools.math_dsl.verify` — re-validates committed `essentials.json`
  * `tools/tests/test_w244_money_collect_kernel.py` — acceptance suite
"""
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class MoneyCollectParams:
    """Closed-form model inputs."""
    p_per_cell: float                # P(money symbol on a grid cell, per spin)
    n_cells: int                     # total grid cells (e.g. 15 for 5×3)
    trigger_count_min: int           # ≥N money on initial spin → trigger
    value_table: dict[float, float]  # {value_x_bet: weight} — unnormalised OK
    respins_reset: int = 3           # respin pool size, default 3
    grid_cap: int | None = None      # absolute cap; defaults to n_cells

    def __post_init__(self):
        if not (0.0 <= self.p_per_cell <= 1.0):
            raise ValueError(f"p_per_cell {self.p_per_cell} outside [0,1]")
        if self.n_cells < 1:
            raise ValueError(f"n_cells {self.n_cells} < 1")
        if self.trigger_count_min < 1:
            raise ValueError("trigger_count_min must be ≥ 1")
        if self.respins_reset < 1:
            raise ValueError("respins_reset must be ≥ 1")
        if not self.value_table:
            raise ValueError("value_table must be non-empty")
        if any(w < 0 for w in self.value_table.values()):
            raise ValueError("value_table weights must be ≥ 0")


def _normalize_value_table(table: dict[float, float]) -> dict[float, float]:
    """Convert {value: weight} → {value: probability}."""
    total = sum(table.values())
    if total <= 0:
        raise ValueError("value_table sum-of-weights must be > 0")
    return {v: w / total for v, w in table.items()}


def expected_money_value(value_table: dict[float, float]) -> float:
    """E[V] where V is one money symbol's value × bet (normalised table)."""
    probs = _normalize_value_table(value_table)
    return sum(v * p for v, p in probs.items())


def initial_trigger_probability(
    p_per_cell: float,
    n_cells: int,
    trigger_count_min: int,
) -> float:
    """P(≥ trigger_count_min money symbols on initial spin).

    Binomial CDF tail: 1 - F(trigger_count_min - 1; n_cells, p_per_cell).
    """
    if trigger_count_min > n_cells:
        return 0.0
    # Binomial PMF via stable iterative product (Python ints — exact).
    # P(X = k) = C(n,k) * p^k * (1-p)^(n-k)
    q = 1.0 - p_per_cell
    cdf_below = 0.0
    # C(n, 0) = 1
    pmf = q ** n_cells  # k = 0 term
    cdf_below += pmf
    for k in range(1, trigger_count_min):
        # PMF(k) / PMF(k-1) = (n-k+1)/k * p/q
        pmf *= (n_cells - k + 1) / k * (p_per_cell / q) if q > 0 else 0.0
        cdf_below += pmf
    return max(0.0, 1.0 - cdf_below)


def expected_episode_total_value(
    params: MoneyCollectParams,
    initial_locked_mean: float | None = None,
) -> float:
    """E[total_money_value_per_episode | trigger].

    Models the (k_locked, respins_remaining) Markov chain. Episode terminates
    when respins_remaining hits 0 OR k_locked equals grid_cap.

    Per respin:
      * `cells_open = n_cells - k_locked`
      * Each open cell lands money with prob `p_per_cell`.
      * Let m = number of new money on this respin (Binomial(cells_open, p)).
      * If m > 0: k_locked += m, respins_remaining = respins_reset (RESET).
      * If m == 0: respins_remaining -= 1.

    Each newly-locked symbol's value is drawn IID from value_table; episode
    value = (k_locked at termination) × E[V].

    Args:
      params: model inputs.
      initial_locked_mean: E[k_locked at start of bonus | trigger].
        If None, defaults to trigger_count_min (conservative — assumes the
        episode starts with the minimum trigger count, which underestimates
        slightly when triggers come with more than the minimum).
    """
    grid_cap = params.grid_cap or params.n_cells
    if initial_locked_mean is None:
        initial_locked_mean = float(params.trigger_count_min)

    # DP over (k_locked, respins_remaining). We compute the EXPECTED
    # number of locked symbols at termination, then multiply by E[V].
    # State value e_state[k][r] = expected k_locked at termination given
    # we're currently at (k, r). Terminal states: r == 0 OR k == grid_cap.
    n_cells = params.n_cells
    R = params.respins_reset
    p = params.p_per_cell

    # Build state table. k ∈ [0, grid_cap], r ∈ [0, R].
    # Iterate by descending r (since transitions only decrease r OR jump up to R).
    # Actually transitions on m>0 jump r → R, on m==0 r → r-1. So we can
    # do fixed-point iteration over the (k,r) grid.
    e = [[0.0] * (R + 1) for _ in range(grid_cap + 1)]

    # Terminal: r == 0 → e[k][0] = k. k == grid_cap → e[grid_cap][r] = grid_cap.
    for k in range(grid_cap + 1):
        e[k][0] = float(k)
    for r in range(R + 1):
        e[grid_cap][r] = float(grid_cap)

    # Fixed-point: iterate until convergence (typically 2-3 sweeps for these
    # parameters since R ≤ 3 and grid_cap ≤ 25).
    for _ in range(50):
        max_delta = 0.0
        for r in range(1, R + 1):
            for k in range(grid_cap):  # k < grid_cap (terminal handled above)
                cells_open = n_cells - k
                if cells_open <= 0:
                    continue
                # E[next state | (k, r)] = sum_m P(m | cells_open, p) * e[k+m][r' (depends on m)]
                # where r' = R if m > 0 else r - 1.
                exp_next = 0.0
                # m = 0 term: P(no money) = (1-p)^cells_open, transition (k, r-1).
                p_zero = (1 - p) ** cells_open
                exp_next += p_zero * e[k][r - 1]
                # m >= 1 terms: build PMF iteratively.
                # PMF(m=1) = cells_open * p * (1-p)^(cells_open-1)
                pmf_m = p_zero  # tracks PMF(m=0)
                q = 1 - p
                for m in range(1, cells_open + 1):
                    # PMF(m) / PMF(m-1) = (cells_open - m + 1)/m * p/q
                    if q == 0:
                        # p == 1 edge case
                        pmf_m = 1.0 if m == cells_open else 0.0
                    else:
                        pmf_m *= (cells_open - m + 1) / m * (p / q)
                    next_k = min(k + m, grid_cap)
                    next_r = R  # money landed → reset
                    exp_next += pmf_m * e[next_k][next_r]
                delta = abs(exp_next - e[k][r])
                if delta > max_delta:
                    max_delta = delta
                e[k][r] = exp_next
        if max_delta < 1e-12:
            break

    # Start state: (initial_locked_mean, R). Linear interpolation if not integer.
    k_lo = int(initial_locked_mean)
    k_hi = min(k_lo + 1, grid_cap)
    frac = initial_locked_mean - k_lo
    expected_k = (1 - frac) * e[k_lo][R] + frac * e[k_hi][R]
    return expected_k * expected_money_value(params.value_table)


def money_collect_rtp_contribution(
    params: MoneyCollectParams,
) -> dict:
    """High-level: total per-spin RTP contribution + intermediate values.

    Returns a dict with `trigger_p`, `expected_value_per_money`,
    `expected_total_per_episode`, `rtp_contribution`, and the input
    snapshot for audit.
    """
    trig_p = initial_trigger_probability(
        params.p_per_cell, params.n_cells, params.trigger_count_min
    )
    e_v = expected_money_value(params.value_table)
    e_total = expected_episode_total_value(params)
    rtp = trig_p * e_total
    return {
        "trigger_p": trig_p,
        "expected_value_per_money": e_v,
        "expected_total_per_episode": e_total,
        "rtp_contribution": rtp,
        "params": {
            "p_per_cell": params.p_per_cell,
            "n_cells": params.n_cells,
            "trigger_count_min": params.trigger_count_min,
            "respins_reset": params.respins_reset,
            "grid_cap": params.grid_cap or params.n_cells,
            "value_table": dict(params.value_table),
        },
    }

File: rust-sim/src/kernels/money_collect.rs
//! W244.10 — money_collect Rust port.
//!
//! Closed-form: Binomial CDF trigger × Markov DP nad (k_locked, respins_remaining).
//! Mirror of `tools/math_dsl/money_collect.py`.

use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MoneyCollectParams {
    pub p_per_cell: f64,
    pub n_cells: u32,
    pub trigger_count_min: u32,
    /// {value_x_bet: weight}
    pub value_table: BTreeMap<String, f64>,
    #[serde(default = "default_respins_reset")]
    pub respins_reset: u32,
    #[serde(default)]
    pub grid_cap: Option<u32>,
}

fn default_respins_reset() -> u32 { 3 }

impl MoneyCollectParams {
    pub fn validate(&self) -> Result<(), String> {
        if !(0.0..=1.0).contains(&self.p_per_cell) {
            return Err(format!("p_per_cell {} outside [0,1]", self.p_per_cell));
        }
        if self.n_cells < 1 {
            return Err("n_cells must be ≥ 1".into());
        }
        if self.trigger_count_min < 1 {
            return Err("trigger_count_min must be ≥ 1".into());
        }
        if self.respins_reset < 1 {
            return Err("respins_reset must be ≥ 1".into());
        }
        if self.value_table.is_empty() {
            return Err("value_table must be non-empty".into());
        }
        for (_, &w) in &self.value_table {
            if w < 0.0 {
                return Err("value_table weights must be ≥ 0".into());
            }
        }
        Ok(())
    }
}

fn parse_value_table(t: &BTreeMap<String, f64>) -> Vec<(f64, f64)> {
    t.iter()
        .map(|(k, v)| (k.parse::<f64>().unwrap_or(0.0), *v))
        .collect()
}

pub fn expected_money_value(t: &BTreeMap<String, f64>) -> f64 {
    let pairs = parse_value_table(t);
    let total: f64 = pairs.iter().map(|(_, w)| *w).sum();
    if total <= 0.0 {
        return 0.0;
    }
    pairs.iter().map(|(v, w)| v * (w / total)).sum()
}

pub fn initial_trigger_probability(
    p_per_cell: f64, n_cells: u32, trigger_count_min: u32,
) -> f64 {
    if trigger_count_min > n_cells {
        return 0.0;
    }
    let q = 1.0 - p_per_cell;
    if q == 0.0 {
        return if trigger_count_min == 0 { 1.0 } else { 1.0 };
    }
    let n = n_cells;
    let mut pmf = q.powi(n as i32);
    let mut cdf_below = pmf;  // k = 0
    for k in 1..trigger_count_min {
        pmf *= ((n - k + 1) as f64) / (k as f64) * (p_per_cell / q);
        cdf_below += pmf;
    }
    (1.0 - cdf_below).max(0.0)
}

pub fn expected_episode_total_value(
    p: &MoneyCollectParams,
    initial_locked_mean_opt: Option<f64>,
) -> f64 {
    let grid_cap = p.grid_cap.unwrap_or(p.n_cells) as usize;
    let initial_locked_mean = initial_locked_mean_opt
        .unwrap_or(p.trigger_count_min as f64);

    let r_max = p.respins_reset as usize;
    let n = p.n_cells as usize;
    let prob = p.p_per_cell;

    // 2D state matrix: e[k][r] for k in [0, grid_cap], r in [0, R]
    let mut e: Vec<Vec<f64>> = vec![vec![0.0; r_max + 1]; grid_cap + 1];
    // Terminal: e[k][0] = k. e[grid_cap][r] = grid_cap.
    for k in 0..=grid_cap {
        e[k][0] = k as f64;
    }
    for r in 0..=r_max {
        e[grid_cap][r] = grid_cap as f64;
    }

    // Fixed-point iteration
    for _ in 0..50 {
        let mut max_delta = 0.0_f64;
        for r in 1..=r_max {
            for k in 0..grid_cap {
                let cells_open = n.saturating_sub(k);
                if cells_open == 0 {
                    continue;
                }
                // Binomial PMF over m new wilds, m=0..cells_open
                let q = 1.0 - prob;
                let pmf = q.powi(cells_open as i32);
                // m=0: stay at k, r→r-1
                let mut exp_next = pmf * e[k][r - 1];
                let mut pmf_m = pmf;
                for m in 1..=cells_open {
                    if q == 0.0 {
                        pmf_m = if m == cells_open { 1.0 } else { 0.0 };
                    } else {
                        pmf_m *= ((cells_open - m + 1) as f64) / (m as f64)
                            * (prob / q);
                    }
                    let next_k = (k + m).min(grid_cap);
                    let next_r = r_max;  // money landed → reset
                    exp_next += pmf_m * e[next_k][next_r];
                }
                let delta = (exp_next - e[k][r]).abs();
                if delta > max_delta {
                    max_delta = delta;
                }
                e[k][r] = exp_next;
            }
        }
        if max_delta < 1e-12 {
            break;
        }
    }

    // Linear interpolation at initial_locked_mean
    let k_lo = initial_locked_mean as usize;
    let k_hi = (k_lo + 1).min(grid_cap);
    let frac = initial_locked_mean - k_lo as f64;
    let expected_k = (1.0 - frac) * e[k_lo][r_max] + frac * e[k_hi][r_max];

    expected_k * expected_money_value(&p.value_table)
}

#[derive(Debug, Serialize)]
pub struct MoneyCollectResult {
    pub trigger_p: f64,
    pub expected_value_per_money: f64,
    pub expected_total_per_episode: f64,
    pub rtp_contribution: f64,
}

pub fn money_collect_rtp_contribution(p: &MoneyCollectParams) -> MoneyCollectResult {
    let trig = initial_trigger_probability(p.p_per_cell, p.n_cells, p.trigger_count_min);
    let e_v = expected_money_value(&p.value_table);
    let e_total = expected_episode_total_value(p, None);
    let rtp = trig * e_total;
    MoneyCollectResult {
        trigger_p: trig,
        expected_value_per_money: e_v,
        expected_total_per_episode: e_total,
        rtp_contribution: rtp,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn ce_like_params() -> MoneyCollectParams {
        let mut vt: BTreeMap<String, f64> = BTreeMap::new();
        vt.insert("1.0".into(), 50.0);
        vt.insert("2.0".into(), 30.0);
        vt.insert("5.0".into(), 15.0);
        vt.insert("10.0".into(), 4.0);
        vt.insert("50.0".into(), 1.0);
        MoneyCollectParams {
            p_per_cell: 0.04,
            n_cells: 15,
            trigger_count_min: 6,
            value_table: vt,
            respins_reset: 3,
            grid_cap: Some(15),
        }
    }

    #[test]
    fn ce_trigger_p_matches_python_within_eps() {
        let p = initial_trigger_probability(0.04, 15, 6);
        // Python reference: 1.499e-5 (from MONEY_COLLECT_KERNEL.json fixture)
        assert!((p - 1.4991e-5).abs() < 1e-7,
                "trigger_p {} not near 1.499e-5", p);
    }

    #[test]
    fn e_v_uniform() {
        let mut t: BTreeMap<String, f64> = BTreeMap::new();
        t.insert("1.0".into(), 1.0);
        t.insert("2.0".into(), 1.0);
        t.insert("4.0".into(), 1.0);
        let e_v = expected_money_value(&t);
        // (1+2+4)/3 = 2.333
        assert!((e_v - 7.0 / 3.0).abs() < 1e-12);
    }

    #[test]
    fn full_kernel_run() {
        let p = ce_like_params();
        let r = money_collect_rtp_contribution(&p);
        assert!(r.trigger_p > 0.0);
        assert!(r.rtp_contribution > 0.0);
        assert!(r.rtp_contribution < 1.0);
    }
}
